app/services/assignment_service.py

```diff
@@ -21,133 +25,150 @@
         course_service = CourseService(self.session)
 
-        assignment = AssignmentModel(
-            id=id,
-            name=name,
-            directory_path=directory_path,
-            # This is relative to directory_path
-            master_notebook_path=f"{ name }.ipynb",
-            max_attempts=max_attempts,
-            available_date=available_date,
-            due_date=due_date,
-            is_published=is_published
-        )
-
-        self.session.add(assignment)
-        self.session.commit()
-
         master_repository_name = await course_service.get_master_repository_name()
         owner = await course_service.get_instructor_gitea_organization_name()
         branch_name = await course_service.get_master_branch_name()
 
-        master_notebook_path = f"{ assignment.directory_path }/{ assignment.master_notebook_path }"
-        # Default empty notebook for JupyterLab 4 w/ Otter Config
-        otter_config_cell = {
-            "cell_type": "raw",
-            "metadata": {},
-            "source": [
-                "# ASSIGNMENT CONFIG\n",
-                "requirements: requirements.txt\n"
-                "export_cell: false\n"
-            ]
-        }
-        title_cell = {
-            "cell_type": "markdown",
-            "metadata": {},
-            "source": [
-                f"# { assignment.name }\n",
-            ]
-        }
-        
-        # See comment above commented out FileOperation below
-        # master_notebook_content = json.dumps({ "cells": [otter_config_cell, title_cell], "metadata": {  "kernelspec": {   "display_name": "Python 3 (ipykernel)",   "language": "python",   "name": "python3"  },  "language_info": {   "codemirror_mode": {    "name": "ipython",    "version": 3   },   "file_extension": ".py",   "mimetype": "text/x-python",   "name": "python",   "nbconvert_exporter": "python",   "pygments_lexer": "ipython3",   "version": "3.11.5"  } }, "nbformat": 4, "nbformat_minor": 5})
-
-        gitignore_path = f"{ directory_path }/.gitignore"
-        gitignore_content = await self.get_gitignore_content(assignment)
-        
-        readme_path = f"{ directory_path }/README.md"
-        readme_content = f"# { name }"
+        assignment_models = []
+        for assignment in assignments:
+            if (
+                assignment.available_date is not None and
+                assignment.due_date is not None and
+                assignment.available_date >= assignment.due_date
+            ):
+                raise AssignmentDueBeforeOpenException
+            
+            try:
+                await self.get_assignment_by_id(assignment.id)
+                raise AssignmentAlreadyExistsException()
+            except AssignmentNotFoundException:
+                pass
+
+            try:
+                await self.get_assignment_by_name(assignment.name)
+                raise AssignmentAlreadyExistsException()
+            except AssignmentNotFoundException:
+                pass
```

Comment:
Why do we need both of these?

Reply:
Actually, why do we need either? I don't think the `createAssignments()` method should be doing implicit validation which requires doing database calls. This method should just be doing what it needs to do to create the actual assignments. There could be another method on the assignment_service that will check the db for if you can create the assignments, but I don't think the `create_assignments()` method should do that implicitly.

Reply:
1. We need both to validate every unique field, not just id or name. I agree it could be a separate method
2. I don't agree with separating creation from validation. It makes more sense in my opinion to raise an "assignment name taken" exception which integrates directly into our FastAPI error handling, than, say, a database integrity violation (which is what you would otherwise get from the method).

Reply:
At the end of the day though, assuming correct usage, there isn't any functional difference when decoupling validation. The only actual difference is assuming misuse (since you could now hit SQLAlchemy errors where not possible before). I'm not entirely opposed to decoupling; it does have its advantages like greater flexibility. On the other hand, it would require a refactor of every service that uses validation logic.

Reply:
> We need to validate every unique field, so both id and name. It needs to be two separate blocks; otherwise, consider if an assignment had a unique ID but non-unique name. It would throw before validating the name. I agree it could be a separate method.

That sounds like a data model problem to me. Just mark the column as UNIQUE then, and if you try and insert something that isn't unique and get an exception back, _then_ you deal with it that way.

Reply:
> I don't agree with separating creation from validation. It makes more sense in my opinion to raise an "assignment name taken" exception which integrates directly into our FastAPI error handling, than, say, a database integrity violation (which is what you would otherwise get from the method).

You don't have to raise a database integrity violation all the way up to FastAPI, you'd catch the database integrity violation and throw your own exception based on what you got. 

I reaaaally think we should separate creation from validation though. We can talk more about why offline.

Reply:
> If you are catching the exception and reraising something more appropriate (NameAlreadyTaken) you are literally just reinventing the already present validation
How are you "reinventing the already present validation"? All I'm saying is that you can catch a database exception and return a FastAPI exception so that the application fails in a graceful way.

> maybe I'm a bit confused why you feel separation of concerns here is so important?
I think in _this instance_ it's just important to separate creation from validation because the validation is requiring doing database calls for every assignment that you pass in to the method. If we can validate all of the assignments in one database call at the beginning that would be better.

Or better yet, I think we should be doing [upserts](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#postgresql-insert-on-conflict) instead of using `add_all()`. Basically just attempt to insert the assignments into the table, but use the `on_conflict_do_nothing()` or `on_conflict_do_update()` method to define what should happen when the assignment already exists in the db.